src/core/include/utils/prng/prng.h

```diff
@@ -0,0 +1,110 @@
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+// OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//==================================================================================
+
+/**
+ * DISCLAIMER: IMPORTANT NOTICE ABOUT FILE MODIFICATIONS
+ * 
+ * This file is used in OpenFHE and ANY PRNG (pseudorandom number generator) OpenFHE uses. 
+ * The file is critical to the functionality of the library.
+ * 
+ * Modifications should only be performed by authorized personnel who understand the potential impacts.
+ * Unauthorized changes may lead to library failure.
+ * 
+ * By proceeding with changes to this file, you acknowledge that you understand the risks involved and
+ * accept full responsibility for any resulting issues.
+ */
+
+/**
+ * Abstract base class for any PRNG engine.
+ */
+
+#ifndef __PRNG_H__
+#define __PRNG_H__
+
+#include <cstdint>
+#include <limits>
+#include <array>
+
+
+// ATTENTION (VERY IMPORTANT):
+//    for any engine class derived from the PRNG class there must be a C function named "createEngineInstance"
+//    returning a dynamically allocated object of that derived class (see how it is done in blake2engine.h) 
+class PRNG {
+public:
+    enum {
+      MAX_SEED_GENS = 16,
+      // the buffer stores 1024 samples of 32-bit integers
+      PRNG_BUFFER_SIZE = 1024
+    };
+
+    // all C++11 distributions used in OpenFHE work with uint32_t by default.
+    // a different data type can be specified if needed for a particular architecture
+    using result_type = uint32_t;
+
+    /**
+     * @brief minimum value used by C++11 distribution generators when no lower
+     * bound is explicitly specified by the user
+     */
+    static constexpr result_type min() {
+      return std::numeric_limits<result_type>::min();
+    }
+
+    /**
+     * @brief maximum value used by C++11 distribution generators when no upper
+     * bound is explicitly specified by the user
+     */
+    static constexpr result_type max() {
+      return std::numeric_limits<result_type>::max();
+    }
+
+    virtual result_type operator()() = 0;
+    virtual ~PRNG() = default;
+
+protected:
+    PRNG() = default;
+    PRNG(const std::array<PRNG::result_type, MAX_SEED_GENS> &seed, PRNG::result_type counter)
+        : m_counter(counter), m_seed(seed) {}
```

Comment:
RESCIND: m_seed is the bulk of the context for the chained hash.  Nevermind.

IGNORE: Suggest immediately calling operator(), and erasing the seed, instead of this being an empty inline.  Even better see comment on m_seed, below.  We should not be keeping copies of seed strewn about.

Reply:
m_seed will go, if the copy constructor goes. will discuss this